src/oauth2.rs

```diff
@@ -1,576 +1,420 @@
         let mut broken_scope = String::from("5 2 4");
-        assert!(SpotifyOAuth::is_scope_subset(
-            &mut needle_scope,
-            &mut haystack_scope
-        ));
-        assert!(!SpotifyOAuth::is_scope_subset(
-            &mut broken_scope,
-            &mut haystack_scope
-        ));
+        assert!(is_scope_subset(&mut needle_scope, &mut haystack_scope));
+        assert!(!is_scope_subset(&mut broken_scope, &mut haystack_scope));
     }
+
     #[test]
-    fn test_save_token_info() {
-        let spotify_oauth = SpotifyOAuth::default()
-            .state(&generate_random_string(16))
-            .scope("playlist-read-private playlist-read-collaborative playlist-modify-public playlist-modify-private streaming ugc-image-upload user-follow-modify user-follow-read user-library-read user-library-modify user-read-private user-read-birthdate user-read-email user-top-read user-read-playback-state user-modify-playback-state user-read-currently-playing user-read-recently-played")
-            .cache_path(PathBuf::from(".spotify_token_cache.json"))
-            .build();
-        let token_info = TokenInfo::default()
+    fn test_write_token() {
+        let tok = TokenBuilder::default()
             .access_token("test-access_token")
-            .token_type("code")
             .expires_in(3600)
             .expires_at(1515841743)
             .scope("playlist-read-private playlist-read-collaborative playlist-modify-public playlist-modify-private streaming ugc-image-upload user-follow-modify user-follow-read user-library-read user-library-modify user-read-private user-read-birthdate user-read-email user-top-read user-read-playback-state user-modify-playback-state user-read-currently-playing user-read-recently-played")
-            .refresh_token("fghjklrftyhujkuiovbnm");
-        match serde_json::to_string(&token_info) {
-            Ok(token_info_string) => {
-                spotify_oauth.save_token_info(&token_info_string);
-                let display = spotify_oauth.cache_path.display();
-                let mut file = match File::open(&spotify_oauth.cache_path) {
-                    Err(why) => panic!("couldn't open {}: {}", display, why.to_string()),
-                    Ok(file) => file,
-                };
-                let mut token_info_string_from_file = String::new();
-                match file.read_to_string(&mut token_info_string_from_file) {
-                    Err(why) => panic!("couldn't read {}: {}", display, why.to_string()),
-                    Ok(_) => {
-                        assert_eq!(token_info_string, token_info_string_from_file);
-                    }
-                }
-            }
-            Err(why) => panic!(
-                "couldn't convert token_info to string: {} ",
-                why.to_string()
-            ),
-        }
-    }
+            .refresh_token("...")
+            .build()
+            .unwrap();
 
-    #[test]
-    fn test_parse_response_code() {
```

Comment:
I wanted to get rid of the `parse_response_code` and just left a TODO for later. But it's relatively easy so I've just made a commit with a solution that is guaranteed to work thanks to the `url` crate. 

The function signature should actually be `ClientResult<Option<String>>` because it may return `ParseError` when the provided URL is wrong, and `None` when the `code` parameter isn't found, but I thought this would be less annoying to deal with. Not sure what you think.

Reply:
> I wanted to improve parse_response_code's implementation to use an external crate and just left a TODO for later.

I think it's not that complex to parse the response code so that we don't need an external crate to parse code. And we still keep the function `parse_response_code`, we should remain its test either, or removing it util introducing an external crate.

> The function signature should actually be ClientResult<Option<String>> because it may return ParseError when the provided URL is wrong, and None when the code parameter isn't found, but I thought this would be less annoying to deal with

Great idea, it makes sense.

Reply:
> I think it's not that complex to parse the response code so that we don't need an external crate to parse code. 

Yes, it's that complex. The previous solution was incorrect. If the code parameter wasn't the first one it wouldn't work at all, for example. We already have the `url` crate for other stuff like `get_authorize_url` so there's no problem using it.

> And we still keep the function parse_response_code, we should remain its test either, or removing it util introducing an

Sure